pkg/urbit/noun/nock.c

```diff
@@ -1665,6 +1672,176 @@ u3n_find(u3_noun key, u3_noun fol)
   return pog_p;
 }
 
+/* _n_prog_free(): free memory retained by program pog_u
+ */
+static void
+_n_prog_free(u3n_prog* pog_u)
+{
+  c3_w i_w;
+  for (i_w = 0; i_w < pog_u->lit_u.len_w; ++i_w) {
+    u3z(pog_u->lit_u.non[i_w]);
+  }
+  for (i_w = 0; i_w < pog_u->mem_u.len_w; ++i_w) {
+    u3z(pog_u->mem_u.sot_u[i_w].key);
+  }
+  for (i_w = 0; i_w < pog_u->cal_u.len_w; ++i_w) {
+    u3j_site_lose(&(pog_u->cal_u.sit_u[i_w]));
+  }
+  for (i_w = 0; i_w < pog_u->reg_u.len_w; ++i_w) {
+    u3j_rite_lose(&(pog_u->reg_u.rit_u[i_w]));
+  }
+  u3a_free(pog_u);
+}
+
+/* _intlen(): find the number of characters the given int
+ *            would take to print.
+ */
+c3_w
+_intlen(c3_w num_w)
+{
+  c3_w len_w=0;
+  while(num_w){
+    num_w/=10;
+    len_w++;
+  }
+  return len_w;
+}
+
+/* _invalid_op(): return true if 'go' is not in 0-2,4
+ *                which means that _slog_bytecode doesn't
+ *                know how to handle it.
+ */
+c3_b
+_invalid_op(c3_y go)
+{
+  return !(go == 0 || go == 1 || go == 2 | go == 4);
+}
+
+/*  _is_pair(): return true if go > 0
+ *              which means the opcode is paired with an argument
+ */
+c3_b
+_is_pair(c3_y go)
+{
+  return (go == 1 || go == 2 | go == 4);
+}
+
+/* _is_indexed(): return true if the opcode is
+ *                in the set of opcodes known to
+ *                use pog_u->lit_u.non
+ */
+c3_b
+_is_indexed(c3_w op)
+{
+  // NOTE: this logic is copied from ___
+  // and must be changed here if that changes.
+  switch (op) {
+  case FIBK: case FIBK+1:
+  case FIBL: case FIBL+1:
+  case LIBK: case LIBK+1:
+  case LIBL: case LIBL+1:
+  case BUSH: case BUSH+1:
+  case SANB: case SANB+1:
+  case KITB: case KITB+1:
+  case MITB: case MITB+1:
+  case HILB: case HILB+1:
+  case HINB: case HINB+1:
+    return 1;
+  default:
+    return 0;
+  }
+}
+
+/* _num_from_pog(): returns an unsigned int of variable size
+ *                  with a side effect of correctly incrementing ip_w
+ *                  defined as a macro to avoid further pointer manipulations
+ */
+#define _num_from_pog(go, pog, ip_w) (\
+  go == 4 ? _n_rewo(pog, &ip_w):      \
+  go == 2 ? _n_resh(pog, &ip_w):      \
+  pog[ip_w++])
+
```

Comment:
@joemfb The goal here was to improve readability and code reuse without having to reason any further with pointer/reference passing or transformation. I also explain this in the in-line docs. 

It was originally just to aid my own reading of _slog_byecode(), and I am happy to rewrite it if a macro here isn't the right way, but I've come to like it.

What do you think?